packages/utilities/src/classNamesFunction.ts

```diff
@@ -1,32 +1,73 @@
+
+const RetVal = '__retval__';
+
 /**
  * Creates a getClassNames function which calls getStyles given the props, and injects them
  * into mergeStyleSets.
+ *
+ * Note that the props you pass in on every render should be in the same order and
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
 export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(): (
   getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
   styleProps?: TStyleProps
 ) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+  let map: Map<any, any> = new Map<any, any>();
 
   const getClassNames = (
     styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
     styleProps: TStyleProps = {} as TStyleProps
   ): IProcessedStyleSet<TStyleSet> => {
-    // styleSet might be undefined if styleFunctionOrObject is undefined, but getStyles should never
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
+    let resultCount = 0;
+
+    for (const propName in styleProps) {
+      if (styleProps.hasOwnProperty(propName)) {
+        let propValue: any = styleProps[propName];
+
+        if (propValue === undefined) {
+          propValue = '__undefined__';
+        }
+        if (propValue === null) {
+          propValue = '__null__';
+        }
+        if (!current.has(propValue)) {
+          current = current.set(propValue, new Map<any, any>());
+        }
+        current = current.get(propValue);
+      }
     }
 
-    const styleSet =
-      styleFunctionOrObject && (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+    if (!(current as any)[RetVal]) {
+      if (styleFunctionOrObject === undefined) {
+        (current as any)[RetVal] = {} as IProcessedStyleSet<TStyleSet>;
+      } else {
+        const styleSet =
+          styleFunctionOrObject &&
+          (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+
+        (current as any)[RetVal] = mergeStyleSets(styleSet as TStyleSet);
+        resultCount++;
+      }
+    }
+
+    if (resultCount > MAX_CACHE_COUNT) {
+      if (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'production') {
+        throw new Error('Styles are being recalculated far too frequently.');
+      }
+      map.clear();
+    }
 
-    return mergeStyleSets(styleSet as TStyleSet);
+    return (current as any)[RetVal] as IProcessedStyleSet<TStyleSet>;
```

Comment:
why not current.get(RetVal)?

Reply:
I will add a comment here, but because RetVal is an attached property; not a key in the hash. :)